### ISSUE TITLE
Is there a script to convert the model to onnx?

### ISSUE THREAD

**Issue by sunmooncode** [princeton-vl/RAFT-Stereo#28]
pth--->onnx

**Comment by lahavlipson** [comment]
We don't have a script to directly convert our model(s) to onnx, however Pytorch has a [tutorial](https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html) to do exactly this.

[torch.onnx.export](https://pytorch.org/docs/master/onnx.html#torch.onnx.export) should be especially helpful.

**Comment by superchenyan** [comment]
import sys
sys.path.append('core')

import argparse
import glob
import numpy as np
import torch
from tqdm import tqdm
from pathlib import Path
from raft_stereo import RAFTStereo
from utils.utils import InputPadder
from PIL import Image
from matplotlib import pyplot as plt

import onnx
from onnxsim import simplify

import ipdb

DEVICE = 'cuda'

def export_onnx(args):
    # ipdb.set_trace()
    model = torch.nn.DataParallel(RAFTStereo(args), device_ids=[0])
    model.load_state_dict(torch.load(args.restore_ckpt))
    model = model.module
    model.to(DEVICE)
    model.eval()

    output_directory = Path(args.output_directory)
    output_directory.mkdir(exist_ok=True)
    
    with torch.no_grad():
        sample_input = (torch.zeros(1, 3, 480, 640).to(DEVICE), torch.zeros(1, 3, 480, 640).to(DEVICE), args.valid_iters, None, True)
        torch.onnx.export(model,                 # model being run
                      sample_input,              # model input (or a tuple for multiple inputs)
                      args.output_directory + "/raft-stereo-realtime.onnx", # where to save the model (can be a file or file-like object)
                      export_params=True,        # store the trained parameter weights inside the model file
                      opset_version=16,          # the ONNX version to export the model to
                      do_constant_folding=True,  # whether to execute constant folding for optimization
                      input_names = ['left', 'right'],   # the model's input names
                      output_names = ['disparity'])
        
        stereo_onnx_model = onnx.load(args.output_directory + "/raft-stereo-realtime.onnx")  # load onnx model
        stereo_model_simp, stereo_check = simplify(stereo_onnx_model)
        assert stereo_check, "Simplified stereo model could not be validated"
        onnx.save(stereo_model_simp, args.output_directory + "/raft_stereo_realtime_simplify.onnx")
        

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--restore_ckpt', help="restore checkpoint", required=True)
    parser.add_argument('--save_numpy', action='store_true', help='save output as numpy arrays')
    parser.add_argument('-l', '--left_imgs', help="path to all first (left) frames", default="datasets/Middlebury/MiddEval3/testH/*/im0.png")
    parser.add_argument('-r', '--right_imgs', help="path to all second (right) frames", default="datasets/Middlebury/MiddEval3/testH/*/im1.png")
    parser.add_argument('--output_directory', help="directory to save output", default="demo_output")
    parser.add_argument('--mixed_precision', action='store_true', help='use mixed precision')
    parser.add_argument('--valid_iters', type=int, default=32, help='number of flow-field updates during forward pass')

    # Architecture choices
    parser.add_argument('--hidden_dims', nargs='+', type=int, default=[128]*3, help="hidden state and context dimensions")
    parser.add_argument('--corr_implementation', choices=["reg", "alt", "reg_cuda", "alt_cuda"], default="reg", help="correlation volume implementation")
    parser.add_argument('--shared_backbone', action='store_true', help="use a single backbone for the context and feature encoders")
    parser.add_argument('--corr_levels', type=int, default=4, help="number of levels in the correlation pyramid")
    parser.add_argument('--corr_radius', type=int, default=4, help="width of the correlation pyramid")
    parser.add_argument('--n_downsample', type=int, default=2, help="resolution of the disparity field (1/2^K)")
    parser.add_argument('--context_norm', type=str, default="batch", choices=['group', 'batch', 'instance', 'none'], help="normalization of context encoder")
    parser.add_argument('--slow_fast_gru', action='store_true', help="iterate the low-res GRUs more frequently")
    parser.add_argument('--n_gru_layers', type=int, default=3, help="number of hidden GRU levels")
    
    args = parser.parse_args()

    export_onnx(args)

**Comment by superchenyan** [comment]
python export_pt2onnx.py --restore_ckpt models/raftstereo-realtime.pth --shared_backbone --n_downsample 3 --n_gru_layers 2 --slow_fast_gru --valid_iters 7 --corr_implementation alt --mixed_precision
